### PR TITLE
#5312: Revert #3580 and try a different workaround (3.7 RC branch)

### PR DESCRIPTION
**Pull request by PhilMiller**
#3580 was an overwrought attempt at working around a bug in NVCC 9.x. On the hypothesis that the original compiler bug was related to re-use/re-assignment of the predicate variable `do_final_reduction`, I tried reverting the workaround and introducing separate variables for the initial and subsequent condition. It seems to have worked for both Clang 10 and NVCC 9.2 and 10.1

Fixes: #5312

### CODE REVIEW

**Comment by PhilMiller** [comment]
Corresponding change to `develop` is in #5325

**Comment by PhilMiller** [comment]
Retest this please

**Review by crtrott — APPROVED** [review]

**Comment by masterleinad** [comment]
The relevant changes wrt. pre-3580 look somewhat like (reverse order)
```diff
diff --git a/core/src/Cuda/Kokkos_Cuda_Parallel_Range.hpp b/core/src/Cuda/Kokkos_Cuda_Parallel_Range.hpp
index 987334300..21ced9551 100644
--- a/core/src/Cuda/Kokkos_Cuda_Parallel_Range.hpp
+++ b/core/src/Cuda/Kokkos_Cuda_Parallel_Range.hpp
@@ -258,12 +258,11 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
 
     // Reduce with final value at blockDim.y - 1 location.
     // Shortcut for length zero reduction
-    bool zero_length        = m_policy.begin() == m_policy.end();
-    bool do_final_reduction = true;
-    if (!zero_length)
+    bool do_final_reduction = m_policy.begin() == m_policy.end();
+    if (!do_final_reduction)
       do_final_reduction = cuda_single_inter_block_reduce_scan<false>(
           final_reducer, blockIdx.x, gridDim.x,
-          kokkos_impl_cuda_shared_memory<word_size_type>(), m_scratch_space,
+          kokkos_impl_cuda_shared_memory<size_type>(), m_scratch_space,
           m_scratch_flags);
 
     if (do_final_reduction) {
diff --git a/core/src/Cuda/Kokkos_Cuda_Parallel_Team.hpp b/core/src/Cuda/Kokkos_Cuda_Parallel_Team.hpp
index cdd16085b..6883aeac4 100644
--- a/core/src/Cuda/Kokkos_Cuda_Parallel_Team.hpp
+++ b/core/src/Cuda/Kokkos_Cuda_Parallel_Team.hpp
@@ -759,15 +759,13 @@ class ParallelReduce<FunctorType, Kokkos::TeamPolicy<Properties...>,
     }
 
     // Reduce with final value at blockDim.y - 1 location.
-    bool zero_length        = m_league_size == 0;
-    bool do_final_reduction = true;
-    if (!zero_length)
-      do_final_reduction = cuda_single_inter_block_reduce_scan<false>(
+    bool do_final_reduce = (m_league_size == 0);
+    if (!do_final_reduce)
+      do_final_reduce = cuda_single_inter_block_reduce_scan<false>(
           final_reducer, blockIdx.x, gridDim.x,
           kokkos_impl_cuda_shared_memory<size_type>(), m_scratch_space,
           m_scratch_flags);
-
-    if (do_final_reduction) {
+    if (do_final_reduce) {
       // This is the final block with the final result at the final threads'
       // location
 ```
I wonder if the switch from `size_type` to `word_size_type` also has some impact still recognizing that we introduced past #3580 for the half type reductions. Still, I wonder if we should change the `Team`  implementation to use `word_size_type`, too.
For comparison:
```
// Conditionally set word_size_type to int16_t or int8_t if value_type is
// smaller than int32_t (Kokkos::Cuda::size_type)
// word_size_type is used to determine the word count, shared memory buffer
// size, and global memory buffer size before the reduction is performed.
// Within the reduction, the word count is recomputed based on word_size_type
// and when calculating indexes into the shared/global memory buffers for
// performing the reduction, word_size_type is used again.
// For scalars > 4 bytes in size, indexing into shared/global memory relies
// on the block and grid dimensions to ensure that we index at the correct
// offset rather than at every 4 byte word; such that, when the join is
// performed, we have the correct data that was copied over in chunks of 4
// bytes.
using word_size_type = std::conditional_t<
      sizeof(value_type) < sizeof(Kokkos::Cuda::size_type),
      std::conditional_t<sizeof(value_type) == 2, int16_t, int8_t>,
      Kokkos::Cuda::size_type>;

using size_type      = Cuda::size_type;
```

**Comment by PhilMiller** [comment]
The OpenMPTarget build timed out. Clearly irrelevant to CUDA backend changes here

**Comment by PhilMiller** [comment]
Comments on the partner PR are maybe worth addressing before merging though

**Comment by crtrott** [comment]
The bug was present before the introduction of word_size_type

**Review by ndellingwood — APPROVED** [review]
Thanks @PhilMiller !

**Comment by PhilMiller** [comment]
I'm going to suggest leaving the `Team` case with `size_type` for the moment, and merging this PR as is, so that we can unblock the release. If we can put together a test that shows a separate failure for that case because of the `size_type/word_size_type` discrepancy, then we should fix that as a separate effort.

**Review by masterleinad — APPROVED** [review]

**Pull request merged**
